### ISSUE TITLE
Open links in external browsers

### ISSUE THREAD

**Issue by Fizzadar** [r0x0r/pywebview#153]
I'm building an email client and need a way to have links open in the OS-level default browser. Thinking that intercepting URL changes before they happen might be the best approach (and using Python's [webbrowser](https://docs.python.org/2/library/webbrowser.html) module to open), preventing the webview from navigating to the new URL (this is how electron does it).

Is there any way to do this with pywebview as-is?

**Comment by shivaprsdv** [comment]
Yes, the method you suggested is definitely doable.

I think the best approach here would be to use new-window requests (using the
normal `<a target="_blank">`) and redirect them to the browser. Currently such
links are stale and don't work in a pywebview window. I remember digging on this
for the old `multiple-windows-support` branch, can't be hard.

Else we'll have to provide some kind of context-menu or keyboard shortcut, both
of which are more complicated.

**Comment by Fizzadar** [comment]
Thanks, @shivaprsdv, I'll look into that branch when I get a chance!

On a sidenote I managed to achieve similar by replacing links to hit the webserver with the original link as a GET parameter, and then use webbrowser to open the link. Capturing new window opening is definitely a neater approach!

**Comment by r0x0r** [comment]
Yeah, you can easily implement this by just routing links to the server and opening a web browser on the Python side.
The `target=_blank` approach is an architectural decision. I reckon we can go forward with it. This way the only way to open a new window will be using pywebview API.

**Comment by shivaprsdv** [comment]
Great. Links to the relevant signals/methods: [GTK](https://webkitgtk.org/reference/webkitgtk/stable/webkitgtk-webkitwebview.html#WebKitWebView-new-window-policy-decision-requested), [Cocoa](https://developer.apple.com/documentation/webkit/webpolicydelegate/1536381-webview?language=objc), [Qt](http://doc.qt.io/archives/qt-5.5/qwebpage.html#acceptNavigationRequest), [Winforms](https://msdn.microsoft.com/en-us/library/aa768337(v=vs.85).aspx)

**Comment by shivaprsdv** [comment]
Pushed proof-of-concept implementation for Mac and GTK. See [`issue153`](https://github.com/r0x0r/pywebview/tree/issue153) branch.

**Comment by r0x0r** [comment]
Windows already opens new links in a stand-alone browser, which happens to be ahem Internet Explorer regardless that the system default browser is. 
WinForms' NewWindow event could otherwise do the job, but unfortunately its arguments do not have a URL of the link. The solution is to capture click events and check if come from their A target=_blank elements. I will get to it at some point.

**Comment by shivaprsdv** [comment]
What about the [NewWindow3](https://msdn.microsoft.com/en-us/library/aa768337(v=vs.85).aspx) event shared above? It has a `bstrUrl` param.
Also the doc above says:
> The NewWindow3 event is only fired when a new instance of Internet Explorer
> is about to be created.
Might be relevant.

**Comment by r0x0r** [comment]
That's a COM event. No idea how to handle COM events from .NET or whether it is possible at all. Sometime I wonder if switching over to .NET was a right decision with the amount of COM fallbacks I have to resort to. Oh well at least coding for .NET is way easier.

**Comment by r0x0r** [comment]
I have implemented Windows support. QT is the only one left to go.

**Comment by r0x0r** [comment]
Implementation for older QT5 versions is pushed as well. I have no idea how to do it with newer (>5.5) QT versions though. The signature of `acceptNavigationRequest` has changed and has no longer necessary information.  `linkClicked` signal is missing as well. 
@shivaprsdv any ideas?

**Comment by shivaprsdv** [comment]
Yeah. It's not new to me. I had met with this ages ago on Cocoa while working
on the old `multiple-windows-support`. I had commented about it ![here](https://github.com/r0x0r/pywebview/issues/43#issuecomment-300197298) and ![here](https://github.com/shivaprsdv/pywebview/commit/d771625b42ac8917b6f3bdb61a1a686803809643#diff-2ec81f56c33197a32ca5ae4c092a6a2aR144).

I gave a try, and the only thing that seems to work (is super hacky, still):
  * createWindow is fired first in 5.5+. It does not have the request or url.
    All we can do is create a dummy page and return it.
  * acceptNavigationRequest is fired as the url loads into the new page. We
    need to intercept it, check if it is coming from our dummy page, retrieve
    the url and load it in the browser. And dispose the dummy page off.

I have confirmed it's working, but still not sure about some bits. Let me see
if there is cleaner way, or I'll push this hack.
